Use FastRandomContext for all tests

sipa opened the pull request:
Currently, the unit tests rely on a mix of an encapsulated FastRandomContext, and some of the Rand*() functions for randomness. This is inefficient, and will become more inefficient when the Rand* functions are changed to always provide strong randomness.

sipa committed:
Add FastRandomContext::rand256() and ::randbytes()

FastRandomContext now provides all functionality that the real Rand* functions
provide.

laanwj commented:
Concept ACK, also been meaning to do this. There is no need for strong randomness in tests.

jnewbery commented:
Tested ACK https://github.com/bitcoin/bitcoin/pull/10321/commits/7666694cfe35ffbfeefcf92c0c36026ee50e6eb0 with one question.

I've completely unscientifically benchmarked this locally by running the tests which were updated (excluding the wallet tests). This seems to marginally speed up test run time:

master:
```
→ for i in {1..10}; do test_bitcoin --run_test=DoS_tests,blockencodings_tests,bloom_tests,checkqueue_tests,coins_tests,dbwrapper_tests,miner_tests,pow_tests,sighash_tests --log_level=test_suite 2>/dev/null | grep "testing time" | cut -d' ' -f7 | rev | cut -c 4- | rev | paste -sd+ - | bc;done
9805962
10284966
9880339
9470773
9775485
10186809
10072789
9522348
9297190
9433797
```
avg: 9.773045.8s

PR10321:
```
→ for i in {1..10}; do test_bitcoin --run_test=DoS_tests,blockencodings_tests,bloom_tests,checkqueue_tests,coins_tests,dbwrapper_tests,miner_tests,pow_tests,sighash_tests --log_level=test_suite 2>/dev/null | grep "testing time" | cut -d' ' -f7 | rev | cut -c 4- | rev | paste -sd+ - | bc;done
9597130
9677749
9644620
9597136
9393690
9542741
9417484
9128269
9600153
9382188
```
avg: 9.498116s

sipa commented on src/random.cpp:
```diff
@@ -247,6 +247,27 @@ void FastRandomContext::RandomSeed()
     requires_seed = false;
 }
 
+uint256 FastRandomContext::rand256()
```
It doesn't feel worthwhile to inline rand256() to me.

TheBlueMatt commented:
Can you scripted-diff this? should be rather trivial to s/GetRandHash/insecure_rand256/?

sipa committed:
Add various insecure_rand wrappers for tests

sipa committed:
Merge test_random.h into test_bitcoin.h

sipa committed:
scripted-diff: Rename cuckoo tests' local rand context

-BEGIN VERIFY SCRIPT-
sed -i 's/insecure_rand/local_rand_ctx/g' src/test/cuckoocache_tests.cpp
-END VERIFY SCRIPT-

sipa committed:
scripted-diff: use insecure_rand256/randrange more

-BEGIN VERIFY SCRIPT-
sed -i "s/\<GetRandHash(/insecure_rand256(/" src/test/*_tests.cpp
sed -i "s/\<GetRand(/insecure_randrange(/" src/test/*_tests.cpp src/test/test_bitcoin.cpp
sed -i 's/\<insecure_rand() % \([0-9]\+\)/insecure_randrange(\1)/g' src/test/*_tests.cpp
-END VERIFY SCRIPT-

sipa committed:
Use randbits instead of ad-hoc emulation in prevector tests

sipa committed:
scripted-diff: Use randbits/bool instead of randrange where possible

-BEGIN VERIFY SCRIPT-
sed -i 's/insecure_randbits(1)/insecure_randbool()/g' src/test/*_tests.cpp
sed -i 's/insecure_randrange(2)/insecure_randbool()/g' src/test/*_tests.cpp
sed -i 's/insecure_randrange(4)/insecure_randbits(2)/g' src/test/*_tests.cpp
sed -i 's/insecure_randrange(32)/insecure_randbits(5)/g' src/test/*_tests.cpp
sed -i 's/insecure_randrange(256)/insecure_randbits(8)/g' src/test/*_tests.cpp
-END VERIFY SCRIPT-

sipa commented:
Significant update: the scope is expanded (some more constructions are replaced with more efficient ones), test_random.h was merged into test_bitcoin.h (why was it separate? the code for both was in test_bitcoin.cpp), and some of the changes are done as scripted diffs.

sipa commented:
Addressed @TheBlueMatt's comments.

sipa commented:
Rebased after merge of #10514.

ryanofsky commented on src/test/prevector_tests.cpp:
```diff
@@ -198,67 +198,65 @@ BOOST_AUTO_TEST_CASE(PrevectorTestInt)
                 test.insert(insecure_randrange(test.size() + 1), insecure_rand());
             }
-            if (test.size() > 0 && ((r >> 2) % 4) == 1) {
+            if (test.size() > 0 && insecure_randbits(2) == 1) {
```
In commit "Use randbits instead of ad-hoc emulation in prevector tests"

Can all these `== 1`, `== 2`, `== 3` checks be replaced with `== 0`? That would seem like a simpler style.

sipa replied:
Let's do that some other time.

ryanofsky reviewed (commented):
utACK b09bd02e3cbbb03758b1467fc2a4bb847142c35d

sipa committed:
Replace more rand() % NUM by randranges

sipa committed:
Replace rand() & ((1 << N) - 1) with randbits(N)

sipa committed:
scripted-diff: Use new naming style for insecure_rand* functions

-BEGIN VERIFY SCRIPT-
sed -i 's/\<insecure_randbits(/InsecureRandBits(/g' src/test/*.cpp src/test/*.h src/wallet/test/*.cpp
sed -i 's/\<insecure_randbool(/InsecureRandBool(/g' src/test/*.cpp src/test/*.h src/wallet/test/*.cpp
sed -i 's/\<insecure_randrange(/InsecureRandRange(/g' src/test/*.cpp src/test/*.h src/wallet/test/*.cpp
sed -i 's/\<insecure_randbytes(/InsecureRandBytes(/g' src/test/*.cpp src/test/*.h src/wallet/test/*.cpp
sed -i 's/\<insecure_rand256(/InsecureRand256(/g' src/test/*.cpp src/test/*.h src/wallet/test/*.cpp
sed -i 's/\<insecure_rand(/InsecureRand32(/g' src/test/*.cpp src/test/*.h src/wallet/test/*.cpp
sed -i 's/\<seed_insecure_rand(/SeedInsecureRand(/g' src/test/*.cpp src/test/*.h src/wallet/test/*.cpp
-END VERIFY SCRIPT-

TheBlueMatt commented:
utACK b09bd02e3cbbb03758b1467fc2a4bb847142c35d

jtimon commented:
Concept ACK

sipa commented on src/test/test_bitcoin.h:
```diff
@@ -28,10 +28,12 @@ static inline void seed_insecure_rand(bool fDeterministic = false)
+static inline uint64_t insecure_randbits(int bits) { return insecure_rand_ctx.randbits(bits); }
+static inline uint64_t insecure_randrange(uint64_t range) { return insecure_rand_ctx.randrange(range); }
+static inline bool insecure_randbool() { return insecure_rand_ctx.randbool(); }
+static inline std::vector<unsigned char> insecure_randbytes(size_t len) { return insecure_rand_ctx.randbytes(len); }
```
Done.

sipa commented on src/test/coins_tests.cpp:
```diff
@@ -155,11 +155,11 @@ BOOST_AUTO_TEST_CASE(coins_cache_simulation_test)
                 newcoin.nHeight = 1;
                 if (insecure_randrange(16) == 0 && coin.IsSpent()) {
-                    newcoin.out.scriptPubKey.assign(1 + (insecure_rand() & 0x3F), OP_RETURN);
+                    newcoin.out.scriptPubKey.assign(1 + insecure_randbits(6), OP_RETURN);
```
Moved to the next commit.

ryanofsky reviewed (commented):
utACK e945848582160b23fa0bc6f797e2bd8ac676ee0e. Changes since last review were moving insecure_randbits changes to appropriate commit, and adding a new scripted diff commit to update style of function names.
